File: 700/729 - My Calendar I.java
import java.util.TreeMap;

class MyCalendar {

    TreeMap<Integer, Integer> map;
    
    public MyCalendar() {
        map = new TreeMap<>();
    }
    
    public boolean book(int start, int end) {
        Map.Entry<Integer, Integer> left = map.floorEntry(start), right = map.floorEntry(end - 1);
        if ((left != null && left.getValue() > start) || (right != null && right.getValue() > start)) {
            return false;
        } else {
            map.put(start, end);
            return true;
        }
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */

File: csharp/729.my-calendar-i.cs
/*
 * @lc app=leetcode id=729 lang=csharp
 *
 * [729] My Calendar I
 */

using System.Collections.Generic;

// @lc code=start
public class MyCalendar {

    private SortedList<int, int> Sorted;

    public MyCalendar() {
        Sorted = new SortedList<int, int>();
        Sorted[int.MinValue] = int.MinValue;
        Sorted[int.MaxValue] = int.MaxValue;
    }
    
    public bool Book(int start, int end) {
        int n = Sorted.Count, left = 0, right = n;
        while (left < right) {
            int mid = (left + right) / 2;
            if (Sorted.Keys[mid] <= start) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if (Sorted.Keys[left] < end || Sorted.Values[left - 1] > start) {
            return false;
        }
        Sorted[start] = end;
        return true;
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * bool param_1 = obj.Book(start,end);
 */
// @lc code=end

